asf_heat_pump_suitability/research/exploratory/block_of_flats_classification_model/build_classification_model_block_of_flats.py

```diff
@@ -0,0 +1,798 @@
+        .then(pl.lit("non-block"))
+        .when(pl.col("label") == "TT")
+        .then(pl.lit("TT"))
+        .otherwise(pl.lit("UL"))
+        .alias("block_of_flats"),
+        pl.col("oct25_building_id").is_duplicated().alias("duplicate"),
+        # Filter to duplicated building IDs (meaning the building has been labelled by two different people)
+    )
+    .filter(pl.col("duplicate"))
+    .group_by("oct25_building_id", maintain_order=True)
+    .agg(
+        # Aggregate the multiple labels and other information into a list per building
+        pl.col("label"),
+        pl.col("block_of_flats"),
+        pl.col("confidence"),
+        pl.col("labeller"),
+        pl.col("url").first(),
+    )
+    .with_columns(
+        # Convert lists to structs and then unnest to create one row per building with information from both labellers
+        pl.col("label").list.to_struct(fields=["sublabel_0", "sublabel_1"]),
+        pl.col("block_of_flats").list.to_struct(fields=["label_0", "label_1"]),
+        pl.col("confidence").list.to_struct(fields=["confidence_0", "confidence_1"]),
+        pl.col("labeller").list.to_struct(fields=["labeller_0", "labeller_1"]),
+    )
+    .unnest(columns=["label", "block_of_flats", "confidence", "labeller"])
+    .with_columns(
+        (pl.col("label_0") == pl.col("label_1")).alias("agree_label"),
+        (pl.col("sublabel_0") == pl.col("sublabel_1")).alias("agree_sublabel"),
+    )
+)
+
+# Building IDs where the labels (of the aggregated four categories) do not match between labellers
+drop_building_ids = double_labelled_df.filter(~pl.col("agree_label"))[
+    "oct25_building_id"
+].unique()
+
+# Building IDs where the labels (of the aggregated four categories) match between labellers
+agree_buildings_ids = double_labelled_df.filter(pl.col("agree_label"))[
+    "oct25_building_id"
+].unique()
+
+
+# Filter to final dataset of labelled data for model
+final_labels_df = (
+    all_labels_df.filter(
+        # Drop buildings where labellers do not agree
+        ~pl.col("oct25_building_id").is_in(drop_building_ids)
+    )
+    .cast({"confidence": pl.Int16}, strict=False)
+    .with_columns(
+        # For buildings where labellers agree, set the confidence to 1, otherwise retain original confidence
+        pl.when(pl.col("oct25_building_id").is_in(agree_buildings_ids))
+        .then(1)
+        .otherwise(pl.col("confidence"))
+        .alias("confidence")
+        # Drop duplicate building IDs - note that this affects the distribution of subclasses within groups
+    )
+    .unique(subset=["oct25_building_id"], keep="any", maintain_order=True)
+    .with_columns(
+        # Convert labels to binary classification for model training
+        pl.when(pl.col("label").is_in(block_of_flats_labels))
+        .then(True)
+        .when(pl.col("label").is_in(non_block_labels))
+        .then(False)
+        .otherwise(None)
+        .alias("block_of_flats")
+    )
+)
+
```

Comment:
I added this to check the distribution including the ULs
```suggestion
final_labels_df["block_of_flats"].value_counts()
```